File: 2022/08/part1.js
const fs = require("fs");
const data = fs.readFileSync("input").toString();
const lines = data.split("\n").filter(Boolean);

const trees = lines.map((row) => row.trim().split("").map(Number));

const outsideVisible = trees.length * 2 + trees[0].length * 2;

let insideVisible = new Set();

for (let rowIdx = 1; rowIdx < trees.length - 2; rowIdx++) {
  const row = trees[rowIdx];

  // Scan left to right
  let minHeight = row[0];
  for (let colIdx = 1; colIdx < row.length - 2; colIdx++) {
    const tree = row[colIdx];
    if (tree > minHeight) {
      if (rowIdx === 1 && colIdx === 65) {
        throw new Error("stop");
      }
      insideVisible.add(`${rowIdx}-${colIdx}-${tree}`);
      minHeight = tree;
    }
  }

  // Scan right to left
  const maxHeight = minHeight;
  minHeight = row[row.length - 1];
  for (let colIdx = row.length - 2; colIdx > 0; colIdx--) {
    const tree = row[colIdx];
    if (tree > minHeight) {
      if (rowIdx === 1 && colIdx === 65) {
        throw new Error("stop");
      }
      insideVisible.add(`${rowIdx}-${colIdx}-${tree}`);
      minHeight = tree;
    }
    // Avoid looping over the rest
    if (tree === maxHeight) {
      break;
    }
  }
}

for (let colIdx = 1; colIdx < trees[0].length - 2; colIdx++) {
  // Scan top to bottom
  let minHeight = trees[0][colIdx];
  for (let rowIdx = 1; rowIdx < trees.length - 2; rowIdx++) {
    const tree = trees[rowIdx][colIdx];
    if (tree > minHeight) {
      insideVisible.add(`${rowIdx}-${colIdx}-${tree}`);
      minHeight = tree;
    }
  }

  // Scan bottom to top
  const maxHeight = minHeight;
  minHeight = trees[trees.length - 1][colIdx];
  for (let rowIdx = trees.length - 2; rowIdx > 0; rowIdx--) {
    const tree = trees[rowIdx][colIdx];
    if (tree > minHeight) {
      if (rowIdx === 1 && colIdx === 65) {
        throw new Error("stop");
      }
      insideVisible.add(`${rowIdx}-${colIdx}-${tree}`);
      minHeight = tree;
    }
    // Avoid looping over the rest
    if (tree === maxHeight) {
      break;
    }
  }
}

// console.log(insideVisible);
console.log(outsideVisible + insideVisible.size);

File: 2022/08/part1.dart
import '../utils.dart';

Future<List<List<int>>> parseInput(String path) async {
  final inputLines = readFile(path);
  final inputLineList = await inputLines.toList();

  return inputLineList.map((line) {
    return line.split('').map((char) => int.parse(char)).toList();
  }).toList();
}

int getVisibleInnerTrees(List<List<int>> input) {
  Set<Coordinate> innerTrees = new Set();

  for (int rowIdx in range(1, input.length - 1)) {
    final row = input[rowIdx];
    int minHeight = row[0];

    // Look left to right
    for (int colIdx in range(1, row.length - 1)) {
      final treeHeight = row[colIdx];
      if (treeHeight > minHeight) {
        minHeight = treeHeight;
        innerTrees.add(Coordinate(colIdx, rowIdx));
      }
    }

    // Look right to left
    final maxHeight = minHeight;
    minHeight = row[row.length - 1];
    for (int colIdx in range(1, row.length - 1).reversed) {
      final treeHeight = row[colIdx];
      if (treeHeight > minHeight) {
        minHeight = treeHeight;
        innerTrees.add(Coordinate(colIdx, rowIdx));
      }
      if (treeHeight == maxHeight) {
        break;
      }
    }
  }

  for (int colIdx in range(1, input[0].length - 1)) {
    // Look top to bottom
    int minHeight = input[0][colIdx];
    for (int rowIdx in range(1, input.length - 1)) {
      final treeHeight = input[rowIdx][colIdx];
      if (treeHeight > minHeight) {
        minHeight = treeHeight;
        innerTrees.add(Coordinate(colIdx, rowIdx));
      }
    }

    // Look bottom to top
    final maxHeight = minHeight;
    minHeight = input[input.length - 1][colIdx];
    for (int rowIdx in range(1, input.length - 1).reversed) {
      final treeHeight = input[rowIdx][colIdx];
      if (treeHeight > minHeight) {
        minHeight = treeHeight;
        innerTrees.add(Coordinate(colIdx, rowIdx));
      }
      if (treeHeight == maxHeight) {
        break;
      }
    }
  }

  return innerTrees.length;
}

void main() async {
  final input = await parseInput('08/input');
  final innerTreeCount = getVisibleInnerTrees(input);
  final output = innerTreeCount + input.length * 2 + (input[0].length - 2) * 2;
  print(output);
}
